source/common/upstream/outlier_detection_impl.cc

```diff
@@ -82,28 +98,111 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
   Http::Code http_code = Http::Code::InternalServerError;
 
   switch (result) {
-  case Result::SUCCESS:
+  case Result::EXT_ORIGIN_REQUEST_SUCCESS:
+  case Result::LOCAL_ORIGIN_CONNECT_SUCCESS:
     http_code = Http::Code::OK;
     break;
-  case Result::TIMEOUT:
+  case Result::LOCAL_ORIGIN_TIMEOUT:
     http_code = Http::Code::GatewayTimeout;
     break;
-  case Result::CONNECT_FAILED:
+  case Result::LOCAL_ORIGIN_CONNECT_FAILED:
     http_code = Http::Code::ServiceUnavailable;
     break;
-  case Result::REQUEST_FAILED:
+  case Result::EXT_ORIGIN_REQUEST_FAILED:
     http_code = Http::Code::InternalServerError;
     break;
-  case Result::SERVER_FAILURE:
-    http_code = Http::Code::ServiceUnavailable;
-    break;
   }
 
   return http_code;
 }
 
-void DetectorHostMonitorImpl::putResult(Result result) {
-  putHttpResponseCode(enumToInt(resultToHttpCode(result)));
+// Method is called by putResult when external and local origin errors
+// are not treated differently. All errors are mapped to HTTP codes.
+// Depending on the value of the parameter *code* the function behaves differently:
+// - if the *code* is not defined, mapping uses resultToHttpCode method to do mapping.
+// - if *code* is zero, no mapping takes place: result is dropped and not reported to outlier
```

Comment:
Per my comment I would like to figure out a better API to avoid this 0 hack. Thoughts?

Reply:
Currently putResult is used in 3 scenarios:
- report event with optional HTTP code. The specified HTTP code is directly fed to outlier detector without any conversions.
- report event without optional HTTP code. OutlierDetector uses legacy code https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.cc#L97-L117 to convert the event to HTTP code based on defaults, which is passed to outlier detector.
- report event, but do not convert it to any HTTP code. This is to catch scenario with router, which must not convert LOCAL_ORIGIN_CONNECT_SUCCESS to HTTP code 200. Otherwise it will collide with the response from the real HTTP server.

I somehow have to indicate not to convert to HTTP code to handle scenario 3. nullopt is already used for scenario 2, what means " I am not giving any code, use defaults for conversion".
I think that the cleanest way would be to always require to specify HTTP code in API. So instead of writing:
`putResult(EXT_ORIGIN_REQUEST_SUCCESS)`
one would have to write
`putResult(EXT_ORIGIN_REQUEST_SUCCESS, absl::optional<uint64_t>(200))`
Then specifying nullopt would mean "Do nothing with HTTP code" to handle scenario 3.

I would have to convert all instances of putResult to add the code - about 8-10 places.

WDYT?

Reply:
Fundamentally I think the issue we have is this API is broken and we are reaching the limits of what we can do with it. Why do we need HTTP codes for non-HTTP paths? Can we have a the following APIs:

`putHttpResult(...)` (for this one just make the caller provide the code each time)
`putExternalOriginResult(...)` (maybe this just takes success/failure like for redis)
`putLocalOriginResult(...)`

And then can't we just do the right thing with them?

Reply:
>  Why do we need HTTP codes for non-HTTP paths?

We need them because of backward compatibility. They were previously mapped to HTTP codes.

> putHttpResult(...) (for this one just make the caller provide the code each time)
> putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
> putLocalOriginResult(...)

There is no problem with implementing/using putHttpResult and putExternalOriginResult. They always eventually go to external origin counters (consecutive_5xx) regardless whether we work in split or non-split mode.

The problem is with putLocalOriginResult. I went through all the cases where LOCAL_ORIGIN event is reported:
- for LOCAL_ORIGIN_TIMEOUT, the code actually instructs to which HTTP code it should be translated: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/router/router.cc#L612-L616, so HTTP code must be passed to outlier detector.
- for LOCAL_ORIGIN_CONNECT_SUCCESS: tcp_proxy wants to translate it to code 200 so it behaves as before my changes:https://github.com/cpakulski/envoy/blob/issue/3643/source/common/tcp_proxy/tcp_proxy.cc#L539-L540. However, router does not want to translate CONNECT_SUCCESS to HTTP code 200 because it will report the real code via putHttpResult. So here to indicate that no conversion should happen, I stick value of zero: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/router/router.cc#L1414-L1419

I can implement 

> putHttpResult(...) (for this one just make the caller provide the code each time)
> putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
> putLocalOriginResult(...)

but putLocalOriginResult must carry HTTP code and somehow I have to indicate not to convert to any code to handle router's case. One way would be to modify include/envoy/http/codes.h and add enum InvalidCode = 0, so we do not hack it explicitly with zero, but use valid enum.

If this sounds good I will implement this today.

Reply:
Sorry this is making my head hurt. Can we somehow unify the TCP proxy and router local origin cases so we don't need the special case, even if it is not explicitly the exact same behavior?

Reply:
I understand that this is mind boggling. It is hard to design nice and elegant API to serve all the cases. There is no way to simply unify TCP cases and router cases. I somehow have to distinguish them. In order to avoid a hack with passing value zero I added another event LOCAL_ORIGIN_CONNECT_SUCCESS_FINAL to indicate that there is no other protocol which will report final error/success.

Reply:
I also experimented with implementing 
```
putHttpResult(...) (for this one just make the caller provide the code each time)
putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
putLocalOriginResult(...)
```
but outlier detector's API is most often invoked through wrappers, so I  would have to implement a wrapped for each of those methods. It very quickly became ugly.